File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let p1 = 0, p2 = 1, curr_max = 0, sum_diff = 0;
    while (p2 < prices.length) {
        if (prices[p2] - prices[p1] < curr_max) {
            p1 = p2;
            p2++;
            curr_max = 0;
        }
        while (prices[p2] - prices[p1] >= curr_max) {
            curr_max = prices[p2] - prices[p1];
            p2++;
        }
        sum_diff += curr_max;
    }
    return sum_diff;
};

File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.java
class Solution {
    public int maxProfit(int[] prices) {
        int valley = 0;
        int peak = 1;
        int sum_profit = 0;
        int curr_diff = 0;

        while (valley < prices.length & peak < prices.length) {
            int v_p = prices[valley];
            int p_p = prices[peak];

            if (p_p - v_p > curr_diff) {
                curr_diff = p_p - v_p;
            } else {
                sum_profit += curr_diff;
                curr_diff = 0;
                valley = peak;
            }

            peak++;
        }

        if (curr_diff > 0 & peak == prices.length) sum_profit += curr_diff;

        return sum_profit;

    }
}